SimTracker/TrackerHitAssociation/plugins/ClusterTPAssociationProducerCUDA.cc

```diff
@@ -0,0 +1,230 @@
+      mapping.insert(std::make_pair(trkid, trackingParticle));
+    }
+  }
+  return mapping;
+}
+
+
+void ClusterTPAssociationProducerCUDA::produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+
+  edm::ESHandle<TrackerGeometry> geom;
+  iSetup.get<TrackerDigiGeometryRecord>().get(geom);
+
+  // Pixel DigiSimLink
+  edm::Handle<edm::DetSetVector<PixelDigiSimLink>> sipixelSimLinks;
+  //  iEvent.getByLabel(_pixelSimLinkSrc, sipixelSimLinks);
+  iEvent.getByToken(sipixelSimLinksToken_, sipixelSimLinks);
+
+  // TrackingParticle
+  edm::Handle<TrackingParticleCollection> TPCollectionH;
+  iEvent.getByToken(trackingParticleToken_, TPCollectionH);
+
+  auto mapping = makeMap(iEvent);
+
+  edm::Handle<CUDAProduct<SiPixelDigisCUDA>> gd;
+  iEvent.getByToken(tGpuDigis, gd);
+  edm::Handle<CUDAProduct<TrackingRecHit2DCUDA>> gh;
+  iEvent.getByToken(tGpuHits, gh);
+
+  CUDAScopedContextProduce ctx{*gd};
+  auto const &gDigis = ctx.get(*gd);
+  auto const &gHits = ctx.get(*gh);
+  auto ndigis = gDigis.nDigis();
+  auto nhits = gHits.nHits();
+
+  std::vector<Clus2TP> digi2tp;
+  digi2tp.push_back({{0, 0, 0, 0, 0, 0, 0}});  // put at 0 0
+  for (auto const &links : *sipixelSimLinks) {
+    DetId detId(links.detId());
+    const GeomDetUnit *genericDet = geom->idToDetUnit(detId);
+    uint32_t gind = genericDet->index();
+    for (auto const &link : links) {
+      if (link.fraction() < 0.5f) {
+        continue;
+      }
+      auto tkid = std::make_pair(link.SimTrackId(), link.eventId());
+      auto ipos = mapping.find(tkid);
+      if (ipos != mapping.end()) {
+        uint32_t pt = 1000 * (*ipos).second->pt();
+        uint32_t eta = 10000 * (*ipos).second->eta();
+        uint32_t z0 = 10000 * (*ipos).second->vz();  // in um
+        uint32_t r0 = 10000 * std::sqrt((*ipos).second->vx() * (*ipos).second->vx() +
+                                        (*ipos).second->vy() * (*ipos).second->vy());  // in um
+        digi2tp.push_back({{gind, uint32_t(link.channel()), (*ipos).second.key(), pt, eta, z0, r0}});
+      }
+    }
+  }
+
+  std::sort(digi2tp.begin(), digi2tp.end());
+
+  ctx.emplace(iEvent, tokenGPUProd_, m_gpuAlgo.makeAsync(gDigis, ndigis, gHits, digi2tp.data(), nhits, digi2tp.size(),ctx.stream()));
+
+}
+
+template <typename T>
+std::vector<std::pair<uint32_t, EncodedEventId>>
+//std::pair<uint32_t, EncodedEventId>
+ClusterTPAssociationProducerCUDA::getSimTrackId(const edm::Handle<edm::DetSetVector<T>> &simLinks,
+                                                 const DetId &detId,
+                                                 uint32_t channel) const {
+  //std::pair<uint32_t, EncodedEventId> simTrkId;
+  std::vector<std::pair<uint32_t, EncodedEventId>> simTrkId;
+  auto isearch = simLinks->find(detId);
+  if (isearch != simLinks->end()) {
+    // Loop over DigiSimLink in this det unit
+    edm::DetSet<T> link_detset = (*isearch);
```

Comment:
Doesn't this lead to copying of the internal `std::vector` of `edm::DetSet`? Ok, the `getSimTrackId()` is like this already in the original
https://github.com/cms-patatrack/cmssw/blob/614ee0b4e0d1d457d6bbf7d51e0799f330fa242b/SimTracker/TrackerHitAssociation/plugins/ClusterTPAssociationProducer.cc#L274

Maybe the two classes could use a common `getSimTrackId()` (the function doesn't use any member data so could be easily shared as a free function)?

Reply:
Let's see how this class evolve. At the moment it does not have clients...
After all its was born as an exercise at  a time when no clients for Hits existed.